src/specs/NativeSafeAreaView.ts

```diff
@@ -1,11 +1,11 @@
 import codegenNativeComponent from 'react-native/Libraries/Utilities/codegenNativeComponent';
 import type { WithDefault } from 'react-native/Libraries/Types/CodegenTypes';
 import type { ViewProps, HostComponent } from 'react-native';
+import { EdgeRecord } from '../SafeArea.types';
 
 export interface NativeProps extends ViewProps {
   mode?: WithDefault<'padding' | 'margin', 'padding'>;
-  // Should be `'top' | 'right' | 'bottom' | 'left'` but that generates invalid code.
-  edges?: readonly string[];
+  edges?: EdgeRecord;
```

Comment:
Yeah you right. I though codegen worked like that but it does not. I've tried with:

```
edges?: Readonly<{
    top?: WithDefault<'off' | 'additive' | 'maximum', 'off'>;
    right?: WithDefault<'off' | 'additive' | 'maximum', 'off'>;
    bottom?: WithDefault<'off' | 'additive' | 'maximum', 'off'>;
    left?: WithDefault<'off' | 'additive' | 'maximum', 'off'>;
  }>;
  ``` 
  
 It seems like codegen follows this but now I get a series of errors in cpp files:
 
 ```
 /Users/nikitashanin/projects/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewShadowNode.cpp:60:38: error: no member named 'end' in 'facebook::react::RNCSafeAreaViewEdgesStruct'
  if (std::find(edges.begin(), edges.end(), "left") != edges.end()) {
```
This is the first time I am trying new architecture and documentation on it is still not very clear. It looks like it still tries to work with `edges` property as if it was an array, but I think I am not supposed to make any changes in these cpp files myself, am I?

Reply:
I think the old type was `string` because otherwise it failed. You might need `{ top: string, left: string ... }`

Reply:
It dies not work this way `{ top: string, left: string ... }` either, same error.

```
5: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':app:buildCMakeDebug[arm64-v8a]'.
> com.android.ide.common.process.ProcessException: ninja: Entering directory `/Users/nikitashanin/projects/react-native-safe-area-context/example/android/app/.cxx/Debug/69p2a451/arm64-v8a'
  [0/2] Re-checking globbed directories...
  [1/2] Re-running CMake...
  -- Configuring done
  -- Generating done
  -- Build files have been written to: /Users/nikitashanin/projects/react-native-safe-area-context/example/android/app/.cxx/Debug/69p2a451/arm64-v8a
  [0/2] Re-checking globbed directories...
  [1/5] Building CXX object safeareacontext_autolinked_build/CMakeFiles/react_codegen_safeareacontext.dir/Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewShadowNode.cpp.o
  FAILED: safeareacontext_autolinked_build/CMakeFiles/react_codegen_safeareacontext.dir/Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewShadowNode.cpp.o 
  /Users/nikitashanin/Library/Android/sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/darwin-x86_64/bin/clang++ --target=aarch64-none-linux-android21 --gcc-toolchain=/Users/nikitashanin/Library/Android/sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/darwin-x86_64 --sysroot=/Users/nikitashanin/Library/Android/sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/darwin-x86_64/sysroot -Dreact_codegen_safeareacontext_EXPORTS -I/Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/android/src/main/jni/. -I/Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/android/src/main/jni/../../../../common/cpp -I/Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/android/src/main/jni/../../../build/generated/source/codegen/jni -I/Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/android/src/main/jni/../../../build/generated/source/codegen/jni/react/renderer/components/safeareacontext -isystem /Users/nikitashanin/.gradle/caches/transforms-3/4223004ae9355336a9d95bfd0b9550ed/transformed/jetified-fbjni-0.3.0/prefab/modules/fbjni/include -isystem /Users/nikitashanin/.gradle/caches/transforms-3/84bec54958abf3a45efcfc6e26585866/transformed/jetified-react-android-0.71.7-debug/prefab/modules/folly_runtime/include -isystem /Users/nikitashanin/.gradle/caches/transforms-3/84bec54958abf3a45efcfc6e26585866/transformed/jetified-react-android-0.71.7-debug/prefab/modules/glog/include -isystem /Users/nikitashanin/.gradle/caches/transforms-3/84bec54958abf3a45efcfc6e26585866/transformed/jetified-react-android-0.71.7-debug/prefab/modules/jsi/include -isystem /Users/nikitashanin/.gradle/caches/transforms-3/84bec54958abf3a45efcfc6e26585866/transformed/jetified-react-android-0.71.7-debug/prefab/modules/react_codegen_rncore/include -isystem /Users/nikitashanin/.gradle/caches/transforms-3/84bec54958abf3a45efcfc6e26585866/transformed/jetified-react-android-0.71.7-debug/prefab/modules/react_debug/include -isystem /Users/nikitashanin/.gradle/caches/transforms-3/84bec54958abf3a45efcfc6e26585866/transformed/jetified-react-android-0.71.7-debug/prefab/modules/react_nativemodule_core/include -isystem /Users/nikitashanin/.gradle/caches/transforms-3/84bec54958abf3a45efcfc6e26585866/transformed/jetified-react-android-0.71.7-debug/prefab/modules/react_render_core/include -isystem /Users/nikitashanin/.gradle/caches/transforms-3/84bec54958abf3a45efcfc6e26585866/transformed/jetified-react-android-0.71.7-debug/prefab/modules/react_render_debug/include -isystem /Users/nikitashanin/.gradle/caches/transforms-3/84bec54958abf3a45efcfc6e26585866/transformed/jetified-react-android-0.71.7-debug/prefab/modules/react_render_graphics/include -isystem /Users/nikitashanin/.gradle/caches/transforms-3/84bec54958abf3a45efcfc6e26585866/transformed/jetified-react-android-0.71.7-debug/prefab/modules/react_render_mapbuffer/include -isystem /Users/nikitashanin/.gradle/caches/transforms-3/84bec54958abf3a45efcfc6e26585866/transformed/jetified-react-android-0.71.7-debug/prefab/modules/rrc_view/include -isystem /Users/nikitashanin/.gradle/caches/transforms-3/84bec54958abf3a45efcfc6e26585866/transformed/jetified-react-android-0.71.7-debug/prefab/modules/turbomodulejsijni/include -isystem /Users/nikitashanin/.gradle/caches/transforms-3/84bec54958abf3a45efcfc6e26585866/transformed/jetified-react-android-0.71.7-debug/prefab/modules/yoga/include -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security   -fno-limit-debug-info  -fPIC -fexceptions -frtti -std=c++17 -Wall -Wpedantic -Wno-gnu-zero-variadic-macro-arguments -DLOG_TAG=\"ReactNative\" -DFOLLY_NO_CONFIG=1 -DFOLLY_HAVE_CLOCK_GETTIME=1 -DFOLLY_USE_LIBCPP=1 -DFOLLY_MOBILE=1 -DFOLLY_HAVE_RECVMMSG=1 -DFOLLY_HAVE_PTHREAD=1 -DFOLLY_HAVE_XSI_STRERROR_R=1 -MD -MT safeareacontext_autolinked_build/CMakeFiles/react_codegen_safeareacontext.dir/Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewShadowNode.cpp.o -MF safeareacontext_autolinked_build/CMakeFiles/react_codegen_safeareacontext.dir/Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewShadowNode.cpp.o.d -o safeareacontext_autolinked_build/CMakeFiles/react_codegen_safeareacontext.dir/Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewShadowNode.cpp.o -c /Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewShadowNode.cpp
  In file included from /Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewShadowNode.cpp:1:
  In file included from /Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewShadowNode.h:6:
  /Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/android/src/main/jni/../../../../common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewState.h:19:7: warning: designated initializers are a C++20 extension [-Wc++20-designator]
        .left = (float)value["left"].getDouble(),
        ^
  /Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewShadowNode.cpp:56:23: error: no member named 'begin' in 'facebook::react::RNCSafeAreaViewEdgesStruct'
    if (std::find(edges.begin(), edges.end(), "top") != edges.end()) {
                  ~~~~~ ^
  /Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewShadowNode.cpp:56:38: error: no member named 'end' in 'facebook::react::RNCSafeAreaViewEdgesStruct'
    if (std::find(edges.begin(), edges.end(), "top") != edges.end()) {
                                 ~~~~~ ^
  /Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewShadowNode.cpp:56:61: error: no member named 'end' in 'facebook::react::RNCSafeAreaViewEdgesStruct'
    if (std::find(edges.begin(), edges.end(), "top") != edges.end()) {
                                                        ~~~~~ ^
  /Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewShadowNode.cpp:60:23: error: no member named 'begin' in 'facebook::react::RNCSafeAreaViewEdgesStruct'
    if (std::find(edges.begin(), edges.end(), "left") != edges.end()) {
                  ~~~~~ ^
  /Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewShadowNode.cpp:60:38: error: no member named 'end' in 'facebook::react::RNCSafeAreaViewEdgesStruct'
    if (std::find(edges.begin(), edges.end(), "left") != edges.end()) {
                                 ~~~~~ ^
  /Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewShadowNode.cpp:60:62: error: no member named 'end' in 'facebook::react::RNCSafeAreaViewEdgesStruct'
    if (std::find(edges.begin(), edges.end(), "left") != edges.end()) {
                                                         ~~~~~ ^
  /Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewShadowNode.cpp:64:23: error: no member named 'begin' in 'facebook::react::RNCSafeAreaViewEdgesStruct'
    if (std::find(edges.begin(), edges.end(), "right") != edges.end()) {
                  ~~~~~ ^
  /Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewShadowNode.cpp:64:38: error: no member named 'end' in 'facebook::react::RNCSafeAreaViewEdgesStruct'
    if (std::find(edges.begin(), edges.end(), "right") != edges.end()) {
                                 ~~~~~ ^
  /Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewShadowNode.cpp:64:63: error: no member named 'end' in 'facebook::react::RNCSafeAreaViewEdgesStruct'
    if (std::find(edges.begin(), edges.end(), "right") != edges.end()) {
                                                          ~~~~~ ^
  /Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewShadowNode.cpp:68:23: error: no member named 'begin' in 'facebook::react::RNCSafeAreaViewEdgesStruct'
    if (std::find(edges.begin(), edges.end(), "bottom") != edges.end()) {
                  ~~~~~ ^
  /Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewShadowNode.cpp:68:38: error: no member named 'end' in 'facebook::react::RNCSafeAreaViewEdgesStruct'
    if (std::find(edges.begin(), edges.end(), "bottom") != edges.end()) {
                                 ~~~~~ ^
  /Users/nikitashanin/projects/react-native-safe-area-context/example/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewShadowNode.cpp:68:64: error: no member named 'end' in 'facebook::react::RNCSafeAreaViewEdgesStruct'
    if (std::find(edges.begin(), edges.end(), "bottom") != edges.end()) {
                                                           ~~~~~ ^
  1 warning and 12 errors generated.
  [2/5] Building CXX object CMakeFiles/appmodules.dir/OnLoad.cpp.o
  [3/5] Building CXX object CMakeFiles/appmodules.dir/Users/nikitashanin/projects/react-native-safe-area-context/example/android/app/build/generated/rncli/src/main/jni/rncli.cpp.o
  ninja: build stopped: subcommand failed.
  
  C++ build system [build] failed while executing:
      /Users/nikitashanin/Library/Android/sdk/cmake/3.18.1/bin/ninja \
        -C \
        /Users/nikitashanin/projects/react-native-safe-area-context/example/android/app/.cxx/Debug/69p2a451/arm64-v8a \
        appmodules \
        react_codegen_rngesturehandler_codegen \
        react_codegen_rnscreens \
        react_codegen_safeareacontext
    from /Users/nikitashanin/projects/react-native-safe-area-context/example/android/app
  CMake Warning at /Users/nikitashanin/Library/Android/sdk/ndk/23.1.7779620/build/cmake/android-legacy.toolchain.cmake:416 (message):
    An old version of CMake is being used that cannot automatically detect
    compiler attributes.  Compiler identification is being bypassed.  Some
    values may be wrong or missing.  Update to CMake 3.19 or newer to use
    CMake's built-in compiler identification.
  Call Stack (most recent call first):
    /Users/nikitashanin/Library/Android/sdk/ndk/23.1.7779620/build/cmake/android.toolchain.cmake:55 (include)
    /Users/nikitashanin/projects/react-native-safe-area-context/example/android/app/.cxx/Debug/69p2a451/arm64-v8a/CMakeFiles/3.18.1-g262b901/CMakeSystem.cmake:6 (include)
    CMakeLists.txt:28 (project)
    ```

Not sure what goes wrong. Unless I am really supposed to make changes to that cpp code myself.

Reply:
This is only the native code, so you could just do `readonly string[]`, and then always give an array with exactly 4 values corresponding to `top`, `right`, `bottom`, `left`. Less clean, but we know that an array of strings does work with fabric

Reply:
Nope can't do. `edges` is not an array of string anymore, it is an object. If I change it here back to `readonly string[];` then Android will not build even for old arch, it will demand `setEdges(view: SafeAreaView!, value: ReadableArray?)` in `SafeAreaViewManager.kt`

I am looking at it